### PR TITLE
Allow keywords in function arguments, (not kwarg only support)

### LINKED ISSUE #235: Support for keyword arguments in function calls
**Is your feature request related to a problem? Please describe.**
`Keyword arguments are not supported yet`

example:
```
...
check_datum(own_addr=own_addr, validator_datum=datum, output_datum=datum_value)
...
```

**Describe the solution you'd like**
I'd like to use keyword arguments when calling a function to have a more structured code.

> [Bug bounty](https://github.com/OpShin/opshin/blob/main/BUG_BOUNTY.md): 2000 ADA

### PR DESCRIPTION
**Pull request by SCMusson**
Potentially fixes #235 

implemented `Typedkeyword` class

added `visit_keyword` method to `AggressiveTypeInferencer`.

Keyword arguments are treated much like args unless in the presence of defaults. This does not support kw_defaults, posonlyargs, or kwargs. Simply allows `func(a,b,c)` to be written as `func(arg1=a, arg2=b, arg3=c)`

### CODE REVIEW

**Comment by SCMusson** [comment]
I'm not convinced the CI failure is due to the changes.

> 
>   ImportError
>   cannot import name 'fs_path_id' from 'virtualenv.info' (/home/travis/virtualenv/python3.8.18/lib/python3.8/site-packages/virtualenv/info.py)
>   at ~/virtualenv/python3.8.18/lib/python3.8/site-packages/virtualenv/discovery/builtin.py:9 in <module>
>         5│ import sys
>         6│ from pathlib import Path
>         7│ from typing import TYPE_CHECKING, Callable
>         8│ 
>     →   9│ from virtualenv.info import IS_WIN, fs_path_id
>        10│ 
>        11│ from .discover import Discover
>        12│ from .py_info import PythonInfo
>        13│ from .py_spec import PythonSpec
> Cannot install docopt.
>   - Installing sortedcontainers (2.4.0)
>   - Installing uplc (0.6.10)
>   - Updating wheel (0.41.2 -> 0.43.0)
> The command "poetry install" failed and exited with 1 during .
> 
>

**Comment by nielstron** [comment]
CI failure is likely due to basing on `main` instead of `dev` (main is currently failing too)

**Comment by nielstron** [comment]
Note that the CI is failing because you are not using the `black` formatter. If you install `pre-commit`, this will be automatically run before commiting code.

**Comment by thaopt** [comment]
hi @nielstron, since kwargs currently not supported, i faced a problem when use hash lib: blake32, i want to pass digest size to args but not able to. plz guilde me if any, ex: def(args*, digest_size) => cannot fill value to digest_size arg

**Comment by nielstron** [comment]
hi @thaopt 
it's not possible to pass a digest size to hash functions in opshin. the reason is that cardano smart contracts don't support this.

File: opshin/type_inference.py [hunk 1]
```diff
@@ -604,6 +605,7 @@ def visit_FunctionDef(self, node: FunctionDef) -> TypedFunctionDef:
         self.exit_scope()
         # We need the function type outside for usage
         self.set_variable_type(node.name, tfd.typ)
+        self.FUNCTION_ARGUMENT_REGISTRY[node.name] = node.args.args
```
Comment (nielstron) [inline opshin/type_inference.py:608]: What happens for scoped overwrites? i.e. the following should be accepted
```
def foo(x:int):
    def foo(y:int):
        return y
    foo(y=1)
foo(x=1)
```

Comment (SCMusson) [reply opshin/type_inference.py:608]: `node.name` contains the unique name in the correct scope, with the original name being located in `node.orig_name`. In your example `foo(x:int)` would be called something like `'foo_0` and `foo(y:int)` would be named `foo_1`. In the `visit_Call` method `node.func.id` contains the name of the correct function taking scope into account.

Comment (nielstron) [reply opshin/type_inference.py:608]: I see, neat trick. Please add a test case for this scenario and check the failing test cases in the CI.

**Comment by nielstron** [comment]
This PR seems on a good track to get accepted (modulo a few failing tests in the CI and a requested test case for nested functions)

One more thing: Does this work for attributes that are functions and builtin functions? I am not even sure if the arguments to these functions have names and this is allowed in native Python but it should be clearly documented whether or not this works (to be clear it's fine if it doesn't if reasonably justified).

**Comment by SCMusson** [comment]
This will only work for User defined functions currently. I can add an assert that checks and warns when keywords are used with functions that are not user defined if you wish. I've added a test that should test for scope issues.

On a side note, I can't reproduce the CI errors on my machine.

 

> __________________________ MiscTest.test_illegal_bind __________________________
> Unexpected success
> ______________________ MiscTest.test_uninitialized_access ______________________
> Unexpected success

These correctly appear as XFails on my machine when running 
`coverage run --source=opshin -m pytest opshin/tests/test_misc.py`

**Comment by SCMusson** [comment]
I'm very confused by these unit tests. I extracted only the tests that failed on CI and appended them to the end of my test_keywords.py tests. Suddenly `Keyword_test.test_uninitialized_access` is failing with Unexpected success but when I place them in their own unittest class (see `MiscTest.test_uninitialized_access` below) It gives me the expected failure. 
When I run the code below with coverage `Keyword_test.test_uninitialized_access` failes and `MiscTest.test_uninitialized_access` is successful despite them being identical tests as far as I can tell.

```import unittest

import hypothesis
from hypothesis import given
from hypothesis import strategies as st



from . import PLUTUS_VM_PROFILE
from .. import prelude, builder, Purpose, PlutusContract
from .utils import eval_uplc_value

hypothesis.settings.load_profile(PLUTUS_VM_PROFILE)





class MiscTest(unittest.TestCase):
    @unittest.expectedFailure
    def test_uninitialized_access(self):
        source_code = """
def validator(_: None) -> int:
    b = 1
    def a(n: int) -> int:
      b += 1
      if b == 2:
        return 0
      else:
        return a(n-1)
    return a(1)
        """
        builder._compile(source_code)

    @unittest.expectedFailure
    def test_illegal_bind(self):
        source_code = """
def validator(_: None) -> int:
    b = 1
    def a(n: int, *args: int) -> int:
      if n == 0:
        return 100
      if b == 2:
        return 0
      b = 2
      return a(n-1)
    return a(2)
        """
        builder._compile(source_code)



class Keyword_tests(unittest.TestCase):
    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_all_keywords(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int, z: int) -> int:
    return (x-y)*z

def validator(a: int, b: int, c: int) -> int:
    return simple_example(x=a, y=b, z=c)
"""
        ret = eval_uplc_value(source_code, x, y, z)
        self.assertEqual(ret, (x - y) * z)

    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_mixture_args_and_keywords(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int, z: int) -> int:
    return (x-y)*z

def validator(a: int, b: int, c: int) -> int:
    return simple_example(a, b, z=c)
"""
        ret = eval_uplc_value(source_code, x, y, z)
        self.assertEqual(ret, (x - y) * z)

    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_keyword_position_independence(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int, z:int) -> int:
    return (x-y)*z

def validator(a: int, b: int, c: int) -> int:
    return simple_example(z=c, x=a, y=b)
"""
        ret = eval_uplc_value(source_code, x, y, z)
        self.assertEqual(ret, (x - y) * z)

    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_arg_after_keyword_failure(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int, z:int) -> int:
    return (x-y)*z

def validator(a: int, b: int, c: int) -> int:
    return simple_example(x=a, y=b, c)
"""
        with self.assertRaises(Exception):
            ret = eval_uplc_value(source_code, x, y, z)

    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_too_many_keywords_failure(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int) -> int:
    return x-y

def validator(a: int, b: int, c: int) -> int:
    return simple_example(x=a, y=b, z=c)
"""
        with self.assertRaises(Exception):
            ret = eval_uplc_value(source_code, x, y, z)
            self.assertEqual(ret, (x * y) + z)

    @given(x=st.integers(), y=st.integers(), z=st.integers())
    def test_incorrect_keywords_failure(self, x: int, y: int, z: int):
        source_code = """
def simple_example(x: int, y: int, z: int) -> int:
    return (x-y)*z

def validator(a: int, b: int, c: int) -> int:
    return simple_example(x=a, y=b, k=c)
"""
        with self.assertRaises(Exception):
            ret = eval_uplc_value(source_code, x, y, z)


    #Why does this one fail but the identical test above doesnt?
    @unittest.expectedFailure
    def test_uninitialized_access(self):
        source_code = """
def validator(_: None) -> int:
    b = 1
    def a(n: int) -> int:
      b += 1
      if b == 2:
        return 0
      else:
        return a(n-1)
    return a(1)
        """
        builder._compile(source_code)

    @unittest.expectedFailure
    def test_illegal_bind(self):
        source_code = """
def validator(_: None) -> int:
    b = 1
    def a(n: int, *args: int) -> int:
      if n == 0:
        return 100
      if b == 2:
        return 0
      b = 2
      return a(n-1)
    return a(2)
        """
        builder._compile(source_code)
```

**Comment by nielstron** [comment]
The function name registry seems to be a default class object with a mutable argument - global side effects are likely. Try initializing it in init or via a default_factory

**Comment by SCMusson** [comment]
I thought that might be the case too (I moved its definition into __init__ in a test) but It did not seem to make a difference although I think I should change the registry to not be a class object regardless. Additionally, I get the same error when running the dev branch (obviously with the extra keyword tests also failing). I'll return to have another look at this later, I'll try and find a minimal code set that reproduces this on my machine and if it fails on a fresh install on the dev branch (I'll give docker a go too) I'll report it in a new issue.

**Comment by SCMusson** [comment]
@nielstron given I could reproduce the error on the dev branch I looked into what was causing it. My tests include a test using the variable `b` and the failing tests are scope tests that also use the variable `b`. I believe the error is not in my changes but because `AggressiveTypeInferencer.scopes` is defined as a default class variable that is mutable. By changing AggresiveTypeInferencer.scopes to be defined in `__init__` as a instance variable i fixed the problem. I'll try pushing it and see if CI is fixed. I don't know if it will break something else, is scopes needed as a class variable?

**Comment by SCMusson** [comment]
All tests passed. I'm guessing that modifying `AggressiveTypeInferencer.scopes` hasn't broken anything then. I'm also guessing there's no bug bounty for finding and fixing a bug while implementing a feature.

**Comment by nielstron** [comment]
Unfortunately there was no bug bounty priced out for this specific bug but thanks for looking into it and resolving it, even though it was not caused by your PR!

Anyways since the tests have passed and the results look good I am officially considering the issue resolved and the bug bounty claimed.

Please do share your ADA address so that I can proceed with the payout :)

File: opshin/tests/test_keywords.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        source_code = """
+def simple_example(x: int, y: int) -> int:
+    return x-y
+
+def validator(a: int, b: int, c: int) -> int:
+    return simple_example(x=a, y=b, z=c)
+"""
+        with self.assertRaises(Exception):
+            ret = eval_uplc_value(source_code, x, y, z)
+            self.assertEqual(ret, (x * y) + z)
```
[the remark this answers is not in the document]
Comment (nielstron) [reply opshin/tests/test_keywords.py:73]: @SCMusson I actually agree with the review bot here, what's the reason there is an assert Equal inside the assert raises?

File: opshin/tests/test_keywords.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    @given(x=st.integers(), y=st.integers(), z=st.integers())
+    def test_correct_scope(self, x: int, y: int, z: int):
+        source_code = """
+def simple_example(x: int, y: int, z: int) -> int:
+    def simple_example(new_x: int, new_z: int) -> int:
+        return new_x-new_z
+    return simple_example(new_x = x, new_z = z) * y
+
+def validatior(a: int, b: int, c: int) -> int:
+    return simple_example(x=a, y=b, k=c)
```
[the remark this answers is not in the document]
Comment (nielstron) [reply opshin/tests/test_keywords.py]: This is correct, please fix the name

File: opshin/tests/test_keywords.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SCMusson) [reply opshin/tests/test_keywords.py]: It's a copy paste line. I've just removed it

**Comment by SCMusson** [comment]
addr1qxsprna8457kszz9kzffg7y886nme4llpnprr6s3pzgeg7d93vtsr3yl4f3553sxfx7wzx58cx5pnhekvtlgqf55lw9s69kv6a

**Comment by nielstron** [comment]
The payment has been initiated 🎉
https://cardanoscan.io/transaction/f22dea4d5a013812e3ba32ccc8c8f96a5d7b9b48d489e0a631fbb6934ecf2aa6

**Comment by nielstron** [comment]
@SCMusson one more question: Does this affect class initializers? I guess they are currently not counted as user defined function? But they should be straightforward to add.

i.e.
```python
class A(PlutusData):
   a: int
   b: int
   
x = A(a=1, b=2) # does this work?
```

**Comment by SCMusson** [comment]
It doesn't affect class initializers. Your example will not work. You'll get an error message like:

`opshin.util.CompilerError: (AssertionError('Keyword arguments can only be used with user defined functions'), <ast.Call object at 0x72a136923b80>, 'Static Type Inference')`

**Pull request merged**